db_store/src/settings.rs

```diff
@@ -1,45 +1,70 @@
-use crate::Result;
+use crate::{iam_auth_pool, Error, Result};
 use serde::Deserialize;
 use sqlx::{
     postgres::{PgConnectOptions, PgPoolOptions},
     Pool, Postgres,
 };
 
+#[derive(Deserialize, Debug, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum AuthType {
+    Postgres,
+    Iam,
+}
+
 #[derive(Debug, Deserialize, Clone)]
 pub struct Settings {
-    /// Max open connections to the database. If absent a default is calculated
-    /// by application code
-    pub max_connections: Option<u32>,
-    /// URL to access the postgres database. For example:
-    /// postgres://postgres:postgres@127.0.0.1:5432/mobile_index_db If the url
-    /// is not specified the following environment variables are used to pick up
-    /// the database settings:
-    ///
-    ///  * `PGHOST`
-    ///  * `PGPORT`
-    ///  * `PGUSER`
-    ///  * `PGPASSWORD`
-    ///  * `PGDATABASE`
-    ///  * `PGSSLROOTCERT`
-    ///  * `PGSSLMODE`
-    ///  * `PGAPPNAME`
-    pub url: Option<String>,
+    pub max_connections: u32,
+
+    pub host: String,
+    pub port: u16,
+    pub database: String,
+    pub username: String,
+
+    #[serde(default = "default_auth_type")]
+    auth_type: AuthType,
+
+    postgres_password: Option<String>,
+
+    pub iam_role_arn: Option<String>,
+    pub iam_role_session_name: Option<String>,
+    pub iam_duration_seconds: Option<i32>,
+    pub iam_region: Option<String>,
+}
+
+fn default_auth_type() -> AuthType {
+    AuthType::Postgres
 }
 
 impl Settings {
-    pub async fn connect(&self, default_max_connections: usize) -> Result<Pool<Postgres>> {
-        let connect_options = if let Some(url) = &self.url {
-            url.parse()?
-        } else {
-            PgConnectOptions::new()
-        };
-        let pool = PgPoolOptions::new()
-            .max_connections(
-                self.max_connections
-                    .unwrap_or(default_max_connections as u32),
-            )
-            .connect_with(connect_options)
-            .await?;
+    pub async fn connect(
+        &self,
+        shutdown: triggered::Listener,
+    ) -> Result<(Pool<Postgres>, futures::future::BoxFuture<'static, Result>)> {
+        match self.auth_type {
+            AuthType::Postgres => match self.simple_connect().await {
+                Ok(pool) => Ok((pool, Box::pin(async move { Ok(()) }))),
```

Comment:
since this empty "handle" task isn't actually doing anything it won't hold up shutting down the app even though it's not listening for a shutdown signal? it'll just return immediately and the `try_join` in an app's main function will just keep on waiting for the _rest_ of the threads to complete or error?

Reply:
This Future will immediately resolve to an `OK(())` Result and not cause the try_join to hang, since it is already resolved.